File: json-schema/research/format_duration/runner.rb
# encoding: UTF-8
# frozen_string_literal: true
#
# Ruby runner for the `duration` format conformance corpus.
#
# Ruby's regex engine is Onigmo. Two pinned-target adjustments matter for the
# fully-anchored pinned duration regex:
#
#   1. `^`/`$` are LINE anchors in Ruby with no non-multiline mode, so the
#      generator emits `\A` (start of string) for `^` and `\z` (end of string,
#      NO trailing-\n exception) for `$`. Without this, Ruby's `$` would accept
#      a trailing "\n" (the `newline-tail` case), diverging from Go/JS. The
#      pinned regex has exactly one leading `^` and one trailing `$`, so the
#      rewrite is exact.
#
#   2. `\d` is ASCII by default in Ruby (verified), so no ASCII-mode flag is
#      needed for the duration regex (it contains no `\b`, so the Onigmo
#      Unicode-`\b` quirk does not apply here).
#
# Runtime matching is unanchored `match?` (the anchors are in the pattern).
#
# Emits JSON Lines to stdout: {"id","engine":"ruby","compiled":bool,"matched":bool|null}
#
# Run: ruby runner.rb [corpus.json]

require "json"

def main
  path = ARGV[0] || "corpus.json"
  corpus = JSON.parse(File.read(path, encoding: "UTF-8"))
  pinned = corpus["pinned_regex"]

  # ^ -> \A , trailing $ -> \z  (the pinned regex's only anchors).
  emitted = pinned.dup
  emitted = "\\A" + emitted[1..] if emitted.start_with?("^")
  emitted = emitted[0..-2] + "\\z" if emitted.end_with?("$")

  compiled = false
  rx = nil
  begin
    rx = Regexp.new(emitted)
    compiled = true
  rescue RegexpError
    compiled = false
  end

  corpus["cases"].each do |c|
    matched = compiled ? rx.match?(c["value"]) : nil
    $stdout.puts JSON.generate("id" => c["id"], "engine" => "ruby",
                               "compiled" => compiled, "matched" => matched)
  end
end

main if __FILE__ == $PROGRAM_NAME

File: json-schema/research/format_duration/runner.py
#!/usr/bin/env python3
"""Python runner for the `duration` format conformance corpus.

Compiles the SINGLE generator-owned pinned regex (from corpus.json's
`pinned_regex`) with re.compile(p, re.ASCII) -- the pinned choice for the
Python target -- and matches each corpus value with re.search (unanchored;
the anchors are in the pattern).

$ NORMALIZATION: Python's `$` matches at end-of-input OR just before a single
trailing `\n`, so a raw `$` would ACCEPT "P1Y\n" (the `newline-tail` case),
diverging from Go/JS. The `pattern` spec pins the fix -- the generator emits
`\Z` (strict end-of-input, no trailing-\n exception) instead of `$` for the
Python target. This runner applies that exact trailing `$`->`\Z` rewrite, so it
tests the FORM the generator would emit. (The pinned regex uses [0-9] rather
than \\d, so re.ASCII is not even load-bearing here, but is kept as the pinned
recipe.)

Reads corpus.json (argv[1] or ./corpus.json) and emits JSON Lines to stdout:
    {"id","engine":"python","compiled":bool,"matched":bool|null}

Run: python3 runner.py [corpus.json]
"""
import json
import re
import sys


def main() -> None:
    path = sys.argv[1] if len(sys.argv) > 1 else "corpus.json"
    with open(path, encoding="utf-8") as fh:
        corpus = json.load(fh)

    pinned = corpus["pinned_regex"]
    # trailing `$` -> `\Z` (strict end-of-input; Python `$` has a trailing-\n exception).
    emitted = pinned[:-1] + r"\Z" if pinned.endswith("$") else pinned

    compiled = False
    rx = None
    try:
        rx = re.compile(emitted, re.ASCII)
        compiled = True
    except re.error:
        compiled = False

    out = sys.stdout
    for k in corpus["cases"]:
        matched = (rx.search(k["value"]) is not None) if compiled else None
        out.write(
            json.dumps(
                {"id": k["id"], "engine": "python", "compiled": compiled, "matched": matched},
                ensure_ascii=False,
            )
            + "\n"
        )


if __name__ == "__main__":
    main()
